src/hardware/innovation.cpp

```diff
@@ -155,68 +157,78 @@ uint8_t Innovation::ReadFromPort(io_port_t port, io_width_t)
 
 void Innovation::WriteToPort(io_port_t port, io_val_t value, io_width_t)
 {
-	const auto now = PIC_FullIndex();
+	RenderUpToNow();
 
-	// Turn on the channel after the data's written
-	if (!is_enabled) {
-		assert(channel);
-		channel->Enable(true);
-		is_enabled = true;
-	} else {
-		RenderForMs(now - last_render_time);
-	}
-	last_render_time = now;
+	unused_for_ms = 0;
 
 	const auto data = check_cast<uint8_t>(value);
 	const auto sid_port = static_cast<io_port_t>(port - base_port);
 	service->write(sid_port, data);
-	unwritten_for_ms = 0;
 }
 
-int16_t Innovation::RenderOnce()
+void Innovation::RenderUpToNow()
 {
-	int16_t sample = 0;
-	while (!service->clock(1, &sample))
-		; // cycle until we have a sample
-	if (!sample) {
-		++silent_frames;
-		return 0;
+	const auto now = PIC_FullIndex();
+
+	// Wake up the channel and update the last rendered time datum.
+	assert(channel);
+	if (!channel->is_enabled) {
+		channel->Enable(true);
+		last_rendered_ms = now;
+		return;
 	}
-	silent_frames = 0;
-	return check_cast<int16_t>(sample * 2);
+	// Keep rendering until we're current
+	for (; last_rendered_ms < now; last_rendered_ms += ms_per_clock)
+		if (float frame = 0.0f; MaybeRenderFrame(frame))
+			fifo.emplace(frame);
 }
 
-void Innovation::RenderForMs(const double duration_ms)
+int16_t Innovation::TallySilence(const int16_t sample)
 {
-	auto render_count = iround(duration_ms * frame_rate_per_ms);
-	while (render_count-- > 0)
-		fifo.push(RenderOnce());
+	if (!sample)
+		++silent_frames;
+	else
+		silent_frames = 0;
+	return sample;
 }
 
-double Innovation::ConvertFramesToMs(const int frames)
+bool Innovation::MaybeRenderFrame(float &frame)
 {
-	return frames / frame_rate_per_ms;
+	assert(service);
+
+	int16_t sample            = {0};
```

Comment:
minor: newline after this?

Reply:
Yeah I tend to agree with that @kcgen. Sometimes I undo the block-style auto-formatted output manually, but overall I think it's still worth it perhaps?

Reply:
yup - I agree; we're not beholden to the tool - so we might as well let it add value when we're OK with it, and then other times use what looks best (if block formatting doesn't work).  :+1: